### ISSUE TITLE
Update epsilon logic to improve numerical stability

### ISSUE THREAD

**Issue by albanD** [pytorch/pytorch#151110]
Many operations in PyTorch use a variety of "epsilon" to ensure numerical stability or avoid infinite value.
This is used in particular for normalization functions batch/rms/layer norm and optimizers.

These epsilons are usually added following paper formula or historical techniques, to improve uniformisation I would suggest we enforce the following rules:
- the epsilon type in user API should always be `Optional[float]`/`std::optional<double>`.
- the default epsilon value should always be  `torch.finfo(torch_tp).eps`/`std::numeric_limits<tp>::epsilon()` with the input type, to ensure we have an epsilon of appropriate value
- epsilon should be applied via `max( . , eps)` and NOT `+ eps`. This has two main advantages: it ensures no numerical change when not close to 0 and it ensures that if the result can never be `0` (see below).

Curious in particular @rgommers if you have opinions on this?

> it ensures that if the result can never be `0`

This can sound like a surprising statement as the value being added is always mathematically guaranteed to be positive. Unfortunately, we've had a couple examples where mathematically true can be broken due to floating point (https://github.com/pytorch/pytorch/issues/29442 would be the example we looked into the most details).
We've also observed very rare NaNs in some very rare cases with some normalization functions which I suspect are also related.

**Comment by rgommers** [comment]
These rules sound very reasonable, and the principles make sense to me. 

> the default epsilon value should always be  `torch.finfo(torch_tp).eps`/`std::numeric_limits<tp>::epsilon()` with the input type, to ensure we have an epsilon of appropriate value

It's possible that this is too small for some applications - you may want to allow a constant times eps or even sqrt(eps) if plain `dtype.eps` doesn't quite work.


It looks like the default values are currently mostly hardcoded constants, e.g.:

- [`LayerNorm`](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html) and [`BatchNorm2d`](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html) default to 1e-5
- [`RMSprop`](https://pytorch.org/docs/stable/generated/torch.optim.RMSprop.html) and [`CosineSimilarity`](https://pytorch.org/docs/stable/generated/torch.nn.CosineSimilarity.html) default to 1e-8
- [`nn.functional.gumbel_softmax`](https://pytorch.org/docs/stable/generated/torch.nn.functional.gumbel_softmax.html) defaults to 1e-10
- `[nn.functional.normalize`](https://pytorch.org/docs/stable/generated/torch.nn.functional.normalize.html) defaults to 1e-12

Some of that looks a bit random, and it seems unlikely that those default values work well for, for example, both `float16` and `float64`. That said, it seems a bit tricky to make any changes here because of potential backwards compatibility impact - are you proposing to make those changes, or only to document the rules and apply them to new functionality?

**Comment by albanD** [comment]
> It's possible that this is too small for some applications - you may want to allow a constant times eps or even sqrt(eps) if plain dtype.eps doesn't quite work.

Ho yes of course, updated the phrasing!

> That said, it seems a bit tricky to make any changes here because of potential backwards compatibility impact - are you proposing to make those changes, or only to document the rules and apply them to new functionality?

Yes I do want to make a push to actually update these values.
We did that exercise for rms_norm and it lead to numerical wins for some users. So I think it is worth the pain.
It will indeed change the numerics so I expect (at least a subset) of these changes is going to be controversial, but I do think it is worth it.

**Comment by mikaylagawarecki** [comment]
> the default epsilon value should always be based off torch.finfo(torch_tp).eps/std::numeric_limits<tp>::epsilon()

Is `torch_tp` the input dtype? What should the epsilon be when the kernel upcasts inputs for intermediate computation and how will we enforce this?

**Comment by albanD** [comment]
> Is torch_tp the input dtype? How does this proposal handle epsilon be when the kernel upcasts inputs for intermediate computation?

Great point, updated the phrasing above. It is indeed not the "input" directly but the dtype of the Tensor that needs to be `>0` (because it's going to be inverted for example).

**Comment by vadimkantorov** [comment]
> * epsilon should be applied via `max( . , eps)` and NOT `+ eps`. This has two main advantages: it ensures no numerical change when not close to 0 and it ensures that if the result can never be `0` (see below).

Might this also change the current op outputs? And even break the numerics (especially in norm functions) for the cases where the user passes explicitly eps

> Yes I do want to make a push to actually update these values.

I think in unsloth experiments sometimes changing these can have bad effect. Are eps saved to state_dict of LayerNorm? (I don't think so) If not - likely users would ask to opt-out of these new defaults for preserving old checkpoint accuracy?

A middle ground might be to keep the old defaults for fp32 and for fp16 flush to `torch.finfo(dtype).eps` if the fp32 default was too small

A related issue:
- https://github.com/pytorch/pytorch/issues/35666

---

But of course if this can be done - it'd be nice...

Another idea might be pulling the trigger for PyTorch 3.0
